ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol/Crypto/HotKey.hs

```diff
@@ -78,37 +80,41 @@ toPeriod HotKey { hkStart = Absolute.KESPeriod lo, hkEvolution } =
 --
 -- When the given KES period is outside the bounds of the 'HotKey', we return
 -- 'Nothing'.
+--
+-- The old 'HotKey' (and an intermediary ones produced while evolving) is
+-- securely erased from memory.
 evolve ::
      forall c m. (IOLike m, Crypto c)
   => Absolute.KESPeriod
   -> HotKey c
   -> m (Maybe (HotKey c))
-evolve targetPeriod hk = do
-    let mNewHotKey = do
-           targetEvolution <- toEvolution hk targetPeriod
-           key' <- go targetEvolution (hkEvolution hk) (hkKey hk)
-           return HotKey {
-               hkStart     = hkStart hk
-             , hkEnd       = hkEnd   hk
-             , hkEvolution = targetEvolution
-             , hkKey       = key'
-             }
-    -- Any stateful code (for instance, running finalizers to clear the
-    -- memory associated with the old key) would happen here or in (a
-    -- monadic) 'go'.
-    return mNewHotKey
+evolve targetPeriod hk
+  | Just targetEvolution <- toEvolution hk targetPeriod
+  = go targetEvolution (hkEvolution hk) (hkKey hk) >>= \case
+      Nothing   -> return Nothing
+      Just key' -> return $ Just HotKey {
```

Comment:
Forget the returned old keys before returning.

Reply:
Important: we never want to have a dangling pointer to a key that is forgotten. Accessing that key might result in a segfault (?) :scream: